docs/source/model/models-from-code.rst

```diff
@@ -0,0 +1,491 @@
+
+Using Models From Code in a Jupyter Notebook
+--------------------------------------------
+
+`Jupyter <https://jupyter.org/>`_ (IPython Notebooks) are a very convenient way to work with AI applications and modeling in general. One slight limitation that they 
+have is in their cell-based execution model. Due to the nature of how they are defined and run, the models from code feature does not directly support defining 
+a notebook as a model. Rather, this feature requires that models are defined as Python scripts (the file extension **must end in '.py'**). 
+
+Fortunately, the folks that maintain the core kernel that Jupyter uses (`IPython <https://ipython.readthedocs.io/en/stable/interactive/magics.html>`_) have created a 
+number of magic commands that are usable within notebooks to enhance the usability of notebooks as a development environment for AI practitioners. One of the most 
+useful magic commands that can be used within any notebook environment that is based upon IPython (``Jupyter``, ``Databricks Notebooks``, etc.) is the ``%%writefile`` command.
+
+The `%%writefile <https://ipython.readthedocs.io/en/stable/interactive/magics.html#cellmagic-writefile>`_ magic command, when written as the first line of a notebook 
+cell, will capture the contents of the cell (not the entire notebook, mind you, only the current cell scope) with the exception of the magic command itself and write 
+those contents to the file that you define. 
+
+For example, running the following in a notebook:
+
+.. code-block:: none
+
+    %%writefile "./hello.py"
+
+    print("hello!")
+
+Will result in a file being created, located in the same directory as your notebook, that contains:
+
+.. code-block:: python
+
+    print("hello!")
+
+
+.. note::
+    There is an optional ``-a`` append command that can be used with the ``%%writefile`` magic command. This option will **append** the cell contents to the file 
+    being targeted for saving the cell contents to. It is **not recommended** to use this option due to the chances of creating difficult-to-debug overrides within 
+    a script that could contain multiple copies of your model definition logic. It is recommended to use the default behavior of ``%%writefile``, which is to overwrite 
+    the local file each time that the cell is executed to ensure that the state of your cell's contents are always reflected in the saved script file.
+
+
+Examples of Using Models From Code
+----------------------------------
+Each of these examples will show usage of the ``%%writefile`` magic command at the top of the script definition cell block in order to simulate defining the model code or other 
+dependencies from within a single notebook. If you are writing your implementations within an IDE or a text editor, do not place this magic command at the top of your 
+script.
+
+.. tabs::
+
+    .. tab:: Simple Example
```

Comment:
super-nit: The tab does not adjust top-margin nicely. I usually add a h3 header to make it look a bit nicer ([ref](https://mlflow.org/docs/latest/llms/tracing/index.html#automatic-tracing))

![Screenshot 2024-08-14 at 11 59 36](https://github.com/user-attachments/assets/c5c63c50-e963-446b-ae3d-335e8330d661)

Maybe this example is long enough to put some headings in between as well.